### PR TITLE
Async programming tutorial.

### PR DESCRIPTION
**Pull request by rohinb2**

### CODE REVIEW

**Comment by rohinb2** [comment]
* **#10** <a href="https://app.graphite.dev/github/pr/run-house/notebooks/10?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rohinb2 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: docs/async.ipynb [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+   "id": "712e5894-a8ff-4663-8f25-35b7a7988b9c",
+   "metadata": {},
+   "source": [
+    "# Asynchronous Programming\n",
+    "\n",
+    "## Background\n",
+    "*Note that this tutorial assumes basic understanding of Runhouse Functions & Modules; it is recommended that you check out [our functions and modules tutorial](https://www.run.house/docs/tutorials/api-modules) before diving into this one.* \n",
+    "\n",
+    "As we've discussed before, once you take a Python function or module and send it to a Runhouse cluster, the cluster holds your resource in memory, and each time that function or module is called by a client, it simply accesses it in memory and calls it. Under the hood, we have a fully asynchronous server running (FastAPI), and a separate process for each environment where your Runhouse resources live. These processes all have their own async event loops, and if you run synchronous functions on Runhouse, they are ran in a separate thread to allow for many concurrent calls to the cluster. **Note that if you are unfamiliar with asynchronous programming in Python, you should just continue using standard, Python sync functions and leave the performance to us**.\n",
```
Comment (carolineechen) [inline docs/async.ipynb:13]:
```suggestion
    "As we've discussed before, once you take a Python function or module and send it to a Runhouse cluster, the cluster holds your resource in memory, and each time that function or module is called by a client, it simply accesses it in memory and calls it. Under the hood, we have a fully asynchronous server running (FastAPI), and a separate process for each environment where your Runhouse resources live. These processes all have their own async event loops, and if you run synchronous functions on Runhouse, they are run in separate threads to allow for many concurrent calls to the cluster. **Note that if you are unfamiliar with asynchronous programming in Python, you should just continue using standard, Python sync functions and leave the performance to us**.\n",
```

**Review by carolineechen — APPROVED** [review]
couple nits, the back and forth references to sync/async and local/remote makes it a little hard to grasp at times imo but overall content looks good

File: docs/async.ipynb [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    "\n",
+    "## Native Async Functions\n",
+    "\n",
+    "But, what if you're writing code that leverages Python's powerful asynchronous functionality? Luckily, we provide rich async support in a variety of ways. First off, any function that is labeled with Python's async keyword will be *executed within the async event loop*, and not within a separate thread. **This means that you should be very careful that you are not running any costly, synchronous code within an async function, to avoid blocking up your the event loop within your environment on the server. Poorly written async functions will not block the entire Runhouse daemon, but will block other functions within the same environment as the user code.** \n",
```
[the remark this answers is not in the document]
Comment (rohinb2) [reply docs/async.ipynb:17]: I think in this case they flow in the paragraph, so we can just leave em bolded for emphasis

[the remark this answers is not in the document]
Comment (rohinb2) [reply docs/async.ipynb:17]: I'm also down to land it and see how it looks and adjsut

**Pull request merged**